uat/custom-components/client-mosquitto-c/src/MqttConnection.cpp

```diff
@@ -0,0 +1,591 @@
+    if (request) {
+        std::shared_ptr<AsyncResult> result(new AsyncResult(MOSQ_ERR_SUCCESS, mid, qos_count, granted_qos, props));
+        request->submitResult(result);
+    }
+}
+
+void MqttConnection::destroyLocked() {
+    if (m_mosq) {
+        mosquitto_destroy(m_mosq);
+        mosquitto_loop_stop(m_mosq, false);
+        m_mosq = NULL;
+    }
+
+    if (!m_ca_file.empty()) {
+        removeFile(m_ca_file);
+        m_ca_file.clear();
+    }
+
+    if (!m_cert_file.empty()) {
+        removeFile(m_cert_file);
+        m_cert_file.clear();
+    }
+
+    if (!m_key_file.empty()) {
+        removeFile(m_key_file);
+    }
+}
+
+void MqttConnection::destroyUnlocked() {
+    std::lock_guard<std::mutex> lk(m_mutex);
+    destroyLocked();
+}
+
+void MqttConnection::removeFile(std::string & file) {
+    if (!file.empty()) {
+        std::remove(file.c_str());
+        file.clear();
+    }
+}
+
+// FIXME: place credentials to temporary files is dangerous
+std::string MqttConnection::saveToTempFile(const std::string & content) {
+    char buffer[1024];
+    char * filename = tmpnam(buffer);                                   // TODO: replace unsafe tmpnam()
+    if (filename) {
+        FILE* fp = fopen(filename, "w");
+        if (fp) {
+            int rc = fputs(content.c_str(), fp);
+            int error = errno;
+            fclose(fp);
+            if (rc == EOF) {
+                throw MqttException("couldn't write to temporary file", error);        // TODO: add filename
+            }
+            return std::string(filename);
+        } else {
+            throw MqttException("couldn't create temporary file for TLS credentials", errno);
+        }
+    } else {
+        throw MqttException("couldn't generate temporary file name for TLS credentials", errno);
+    }
+}
+
+void MqttConnection::stateCheck() {
+    if (!m_mosq) {
+        throw MqttException("MQTT client is not in connected state", -1);
+    }
+}
+
+
+PendingRequest * MqttConnection::createPendingRequestLocked(int request_id) {
+    PendingRequest * request = new PendingRequest();
+    auto it = m_requests.find(request_id);
+    // remove old
+    if (it != m_requests.end()) {
+        delete it->second;
+    }
+    // insert new
+    m_requests[request_id] = request;
+    return request;
+}
+
+PendingRequest * MqttConnection::getValidPendingRequestLocked(int request_id) {
+    PendingRequest * request = NULL;
+
+    auto it = m_requests.find(request_id);
+    if (it != m_requests.end() && it->second->isValid()) {
+        request = it->second;
+    }
+
+    return request;
+}
+
+void MqttConnection::removePendingRequestUnlocked(int request_id) {
+    std::lock_guard<std::mutex> lk(m_mutex);
+
+    removePendingRequestLocked(request_id);
+}
+
+void MqttConnection::removePendingRequestLocked(int request_id) {
+
+    auto it = m_requests.find(request_id);
+    if (it != m_requests.end()) {
+        delete it->second;
+        m_requests.erase(it);
+    }
+}
+
+/*
```

Comment:
remove unused?

Reply:
removed in 8106e75fbe54a32c4b22de1157aadbf12ffe3a5b